americangut/question/ag_question.py

```diff
@@ -0,0 +1,117 @@
+        if not isinstance(description, str):
+            raise TypeError('description must be a string')
+        if not inspect.isclass(dtype):
+            raise TypeError('dtype must be a class')
+        if not isinstance(clean_name, str) and clean_name is not None:
+            raise TypeError('If supplied, clean_name must be a string')
+
+        # Handles the main information about the data
+        self.name = name
+        self.description = description
+        self.dtype = dtype
+
+        self.type = 'Question'
+        if clean_name is None:
+            self.clean_name = name.replace('_', ' ').title()
+        else:
+            self.clean_name = clean_name
+
+        # Sets up
+        self.free_response = kwargs.get('free_response', False)
+        self.mimmarks = kwargs.get('mimmarks', False)
+        self.ontology = kwargs.get('ontology', None)
+        self.remap_ = kwargs.get('remap', None)
+
+    def check_map(self, map_):
+        """Checks the group exists in the metadata
+
+        Parameters
+        ----------
+        map_ : DataFrame
+            A pandas object containing the data to be analyzed. The
+            Question `name` should be a column in the `map_`.
+
+        """
+        if self.name not in map_.columns:
+            raise ValueError('%s is not a column in the supplied map!'
+                             % self.name)
+
+    def remap_data_type(self, map_, watch=True):
+        """Makes sure the target column in map_ has the correct datatype
+
+        map_ : DataFrame
+            A pandas dataframe containing the column described by the question
+            name.
+        watch : bool
+            A flag to indicate the change should be logged. Should generally
+            be true, unless `remap_data_type` is called before another
+            function is executed.
+
+        """
+        if self.dtype == bool:
+            if not (set(map_[self.name].dropna()).issubset(
+                    self.true_values.union(self.false_values))):
+                raise TypeError('%s cannot be cast to a bool value.'
+                                % self.name)
+
+            def remap_(x):
+                if isinstance(x, str) and x.lower() in self.true_values:
+                    return True
+                elif isinstance(x, str) and x.lower() in self.false_values:
+                    return False
+                elif np.isnan(x):
+                    return x
+                else:
+                    try:
+                        return bool(x)
+                    except:
+                        return np.nan
+        else:
+            def remap_(x):
+                return self.dtype(x)
+
+        map_[self.name] = map_[self.name].apply(remap_).astype(self.dtype)
+        map_.replace('nan', np.nan, inplace=True)
+
+        if watch and self.type in {'Categorical', 'Multiple', 'Clinical',
+                                   'Bool', 'Frequency'}:
+            self._update_order(remap_)
```

Comment:
I'm really confused, as well. Because the issue here has less to do with the `dtype` parameter, and more to do with a difference between the two inherieted classes and the fact that I'm not sure you can modify a function in an inherited class and keep the same API.

There are two main types of questions: `AgCategorical` and `AgContinuous`. These are both derived from `AgQuestion`. Categorical questions track the groups in the data, using the `order` and `extremes` parameter. These are dynamic parameters, which are updated in conjunction with the data in the map. The `_update_order` parameter does nothing to the `AgQuestion.dtype`. It keeps the `order` and `extremes` up to date with the data in the map.